### ISSUE TITLE
function declarations inside `try` blocks have inconsistent behaviour

### ISSUE THREAD

**Issue by cstjean** [JuliaLang/julia#34910]
This code fails:

```julia
julia> try 
           function bar end
           g(::typeof(bar)) = 934
           bar(a) = a+2
       finally end
ERROR: UndefVarError: bar not defined
Stacktrace:
 [1] top-level scope at REPL[3]:3
```

However, commenting either the second or third line of the try block makes it succeed. It seems to me that either `function bar end` should be an error outside of the top-level, or the above code should work... FWIW it came about because I have a macro that works at the top-level, but fails inside a `@testset`. I can pull it out, but it made sense to have it inside.

**Comment by brenhinkeller** [comment]
Not sure what the best label for this is, but does still reproduce in 1.8.3 and 1.9.0-alpha1

**Comment by cstjean** [comment]
Yes, still the same behaviour on 1.8. It's very much a corner case. I don't know what the outcome should be TBH...

**Comment by LilithHafner** [comment]
This is even worse with conditional function declarations
```julia
julia> (let
           if true
               f() = 1
           else
               f() = 2
           end
       end)()
WARNING: Method definition f() in module Main at REPL[66]:3 overwritten at REPL[66]:5.
2

julia> (let
           if false
               f() = 1
           else
               f() = 2
           end
       end)()
WARNING: Method definition f() in module Main at REPL[67]:3 overwritten at REPL[67]:5.
ERROR: UndefVarError: `f` not defined
Stacktrace:
 [1] top-level scope
   @ REPL[67]:5
```
from https://github.com/JuliaLang/julia/issues/5148#issuecomment-1336341977

**Comment by elextr** [comment]
For the first conditional case, IIUC `if` does not create a scope, so doesn't that mean the two definitions are both actually in the `let` scope hence the "overwritten" message?

**Comment by LilithHafner** [comment]
They are in the same scope, but code in the else branch of an `if true` statement should not be executed. For example, at the toplevel, we have this more intuitive behavior:
```julia
julia> if true
           f() = 1
       else
           f() = 2
       end
f (generic function with 1 method)

julia> f()
1

**Comment by elextr** [comment]
On thinking about it some more, and based on my understanding as expressed on #47733 declarations happen irrespective of execution, a declaration after a return still exists throughout the scope even though its never executed.  So that applies to function declarations/method definitions as well.  

So the `if` is not controlling the declaration of the function `f()` at line 3 or the redefinition of the method at line 5, they are not executable statements and you get a redefinition message in both `if true` and `if false` cases.  

But what the `if` is determining is the return value, and in the `true` that is a function object that was declared there, and when called the redefined method is used and returns 2. In the `false` clause its a method redefinition.  I am not sure what that means in terms of the object returned, but it does seem to confuse things.

**Comment by elextr** [comment]
Our posts overlapped, but I understand that the rules of REPL globals are different because its not really a "scope".

**Comment by uniment** [comment]
Also from [#5148 (comment)](https://github.com/JuliaLang/julia/issues/5148#issuecomment-1336341977): 

```julia
julia> function fn_generator3()
                  if rand(Bool)
                      f() = 0
                  else
                      f(x, y) = x+y
                  end
              end
fn_generator3 (generic function with 1 method)

julia> fn_generator3()
(::var"#f#4") (generic function with 2 methods)

julia> fn_generator3()
ERROR: UndefVarError: f not defined

julia> fn_generator3()
(::var"#f#4") (generic function with 2 methods)
```

and 

```julia
julia> i=3
3

julia> function f1()
           i=5; g() = 1
           for i=1:3
               g() = @show(i); g() # ok so we've defined `i` at local scope, parent scope, and global
           end
           g
       end
WARNING: Method definition g() in module Main at REPL[126]:2 overwritten at REPL[126]:4.
f1 (generic function with 1 method)

julia> f1()() # where is `i` *not* defined!?
ERROR: UndefVarError: i not defined
```

As suggested in the [#5148 comment](https://github.com/JuliaLang/julia/issues/5148#issuecomment-1336341977), I think the solution is to throw errors on `const` and named function declarations in conditionally-evaluated code in local scopes (unless nested in an even more-local scope than the conditional), and for any `const` or named function declaration in a local scope to *always* be at the most-local scope, i.e. any local `const` should behave like `local const`, local `g()=...` should behave like `local g()=...`, and local `function f end` should behave like `local function f end`. Any desire to capture and manipulate an identifier at parent scope can be achieved anyway with anonymous functions.

**Comment by uniment** [comment]
I just realized that `local function f end` causes a `malformed expression` syntax error, and I have no idea why that's the case when `local function f() end` is fine 😅 

The OP case seems to do with declaring local functions that type-specialize on locally-defined types. For example:

```julia
julia> b=try
           bar = "Hello, world!"
           h = (::typeof(bar)) -> 934
       finally end
ERROR: UndefVarError: `bar` not defined
```

**Comment by uniment** [comment]
Meanwhile:

```julia
julia> function h end
h (generic function with 0 methods)

julia> b=try
           bar = "hello, world!"
           h(::typeof(bar)) = 934
       finally end
h (generic function with 1 method)

julia> h("hi")
934
```
